### DISCUSSION TITLE
Grasshopper Academy

### DISCUSSION THREAD

**Discussion by coderextreme** [christoph-v/spark#11] [Show and tell]
Grasshopper Academy was a peer-to-peer (maybe one-to-many from teacher to students, then a 1-to-1 from student to teacher) video chat. It was only tested in a peer-to-peer mode.  It was built with Angular 2.0.0 alpha, which is no longer available??   I’ve been trying to resurrect it using Angular 13.

Difference between node.js and Angular.  Angular still uses package.json so npm, yarn, etc are required.  From the Angular app generator, you get a whole website which you customize.   Two important commands in Anugular, namely ‘’’ng build’’’ and ‘’’ng serve’’’ the serve command replaces running an express server with node.js.   The build command converts TypeScript to JavaScript.  Also with node.js one would do rollup, webpack o, or roll your own node.js server.

**Comment by coderextreme** [comment]
I believe I ran Grasshopper Academy without ng serve, probably just ran ng build and deployed to webserver…no ng serve or node.js required for a production server.

**Comment by christoph-v** [comment]
Hi @coderextreme 
Woke up and had an idea about how to answer at this thread.

I think, we will be able to learn a lot from Grasshopper Academy. Please feel free to create a developer branch built on freshest master  branch.
Then create a directory within 3rd-party-input and add the Grasshopper files and directories. Push the developer branch and let me have a look, before I merge your developer branch into master.

Anyway, I would have a few questions, which might be kind of demotivating you, hope it is OK to ask them:
- what would be the advantage for the spark project, if it was using the Angular software?
- Could we use code snippets from the Grasshopper Academy in the spark software? Are you the holder of the original copyright? Do we need to consider someone else's copyrights?
- Could we learn from the Grasshopper Academy software, by having a look to the codes? (Me personally, I need to learn nearly EVERYTHING about Web technology, but I am also talking about potential other - younger - collaborators)
- Please do not forget: the current approach of spark: use plain WebRTC + plain Node.js (for AJAX and SSE) to implement the ALP. This approach does not (yet) comprise the Angular software. 

Good night, again :-)
Yours Christoph

**Reply by coderextreme** [reply]
I committed in the master branch.  We can delete the code and reinstall it in a branch. 
https://github.com/christoph-v/spark/tree/master/3rd-party-input/yottzumm%40gmail.com/grasshopper

There are around 3 errors left before we can start testing.

Advantages to using Angular; It's written in TypeScript which may be a disadvantage when building with peerjs.
Yes, you can use code snippets in the grasshopper/src/app/_foo_.component.ts folder.  The rest is pretty boilerplate and a lot like other Angular apps. It was created with an app generator.  There's lots of copyrighted code in the node_modules folder, especially Angular itself.  I'm not sure what we would do if Google comes after us. I think there's a way to get license reports.

I think you would be most interested in project.component.ts (getUserMedia) and the peerjs stuff (later is not mine).

**Reply by coderextreme** [reply]
I've forgotten where I got me.mov from.  It can be replaced.

**Reply by christoph-v** [reply]
Good evening John,

Thank you for all your efforts with making the Grasshopper Academy files run at the spark project.

I hope I will have some time on one of the next weekends, then I will try to get familiar with the Grasshopper Academy.

Maybe we can take advantage of the Angular software, even when creating an experimental application for spark (the so-called "Open Source Conferencing Tool").

Have a nice week
CP

**Comment by coderextreme** [comment]
I think that Svelte would be a bigger win than angular for speed of
development.   I’d like to stay away from React and Next.js.


Would we be using these inside VS Code?


There used to be a report on all licenses in your node_modules folder.  Hmm

I don’t know who holds the license to the boilerplate code.

Hmm

**Comment by coderextreme** [comment]
Also consider looking at call.component.ts in grasshopper/src/apps

**Reply by coderextreme** [reply]
Various licenses meta data can be found by searching package-lock.json for “license”. https://github.com/christoph-v/spark/blob/master/3rd-party-input/yottzumm%40gmail.com/grasshopper/package-lock.json
